File: ddiapi-client/src/main/java/com/kynetics/updatefactory/ddiclient/api/security/HawkbitAuthenticationRequestInterceptor.java
/*
 * Copyright © 2017-2019 Kynetics LLC
 *
 *  All rights reserved. This program and the accompanying materials
 *  are made available under the terms of the Eclipse Public License v1.0
 *  which accompanies this distribution, and is available at
 *  http://www.eclipse.org/legal/epl-v10.html
 *
 */

package com.kynetics.updatefactory.ddiclient.api.security;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Set;

/**
 * @author Daniele Sergio
 */
public class HawkbitAuthenticationRequestInterceptor implements Interceptor {

    public HawkbitAuthenticationRequestInterceptor(Set<Authentication> authentications) {
        Objects.requireNonNull(authentications);
        this.authentications = new ArrayList<>(authentications);
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        if(authentications.size() == 0){
            return chain.proceed(chain.request());
        }

        final Request originalRequest = chain.request();

        final Request.Builder builder = originalRequest.newBuilder();
        final int size = authentications.size();
        final int exitValue = authenticationUse;
        Response response = null;
        do{
            final Authentication authentication = authentications.get(authenticationUse);
            builder.header(authentication.getHeader(), authentication.getHeaderValue());
            response = chain.proceed(builder.build());
            if(response.code() != 401){
                break;
            }
            authenticationUse = ++authenticationUse % size;
        }while(authenticationUse != exitValue);

        return response;
    }



    private final List<Authentication> authentications;
    private int authenticationUse = 0;
}

File: ddiapi-client/src/main/kotlin/com/kynetics/updatefactory/ddiapiclient/security/HawkbitAuthenticationRequestInterceptor.kt
/*
 * Copyright © 2017-2021  Kynetics  LLC
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package com.kynetics.updatefactory.ddiapiclient.security

import java.io.IOException
import java.util.ArrayList
import java.util.Objects
import okhttp3.Interceptor
import okhttp3.Response

/**
 * @author Daniele Sergio
 */
class HawkbitAuthenticationRequestInterceptor(authentications: Set<Authentication>) : Interceptor {

    private val authentications: List<Authentication>
    private var authenticationUse = 0

    init {
        Objects.requireNonNull<Set<Authentication>>(authentications)
        this.authentications = ArrayList(authentications)
    }

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        if (authentications.isEmpty()) {
            return chain.proceed(chain.request())
        }

        val originalRequest = chain.request()

        val builder = originalRequest.newBuilder()
        val size = authentications.size
        val exitValue = authenticationUse
        var response: Response
        do {
            val authentication = authentications[authenticationUse]
            builder.header(authentication.header, authentication.headerValue)
            response = chain.proceed(builder.build())
            if (response.code() != 401) {
                break
            }
            authenticationUse = ++authenticationUse % size
        } while (authenticationUse != exitValue)

        return response
    }
}
